README.md
# `asgardex-ethereum`

## Modules

- `client` - Custom client for communicating with Ethereum by using [`ethers`](https://github.com/ethers-io/ethers.js)


CHANGELOG.md
# v.0.2.0 (2020-10-06)

-## Add:

- `setVault`
- `vaultTx`
- `normalTx`
- `erc20Tx`
- `estimateNormalTx`
- `estimateGasERC20Tx`
- `getERC20Balance`
- `getTransactionCount`
- Optional `vault?: string` parameter of constructor

## Breaking changes:

- Remove `get address` in favour of `getAddress`
- Remove `get balance()` in favour of using `getBalance` only
- Remove `init()`
- `ethers` has been moved from `dependencies` to `peerDependencies`

## Misc.:

- Update for using latest `ethers@5x`

# v.0.1.0 (2020-05-28)

First release


lib/src/client.d.ts
import { ethers } from 'ethers';
import { Provider, TransactionResponse } from '@ethersproject/abstract-provider';
import { EtherscanProvider } from '@ethersproject/providers';
import { Network, Address, Phrase, NormalTxOpts, Erc20TxOpts, EstimateGasERC20Opts } from './types';
/**
 * Interface for custom Ethereum client
 */
export interface EthereumClient {
    setNetwork(network: Network): Network;
    setPhrase(phrase?: string): void;
    getAddress(): string;
    getBalance(address: Address): Promise<ethers.BigNumberish>;
    getBlockNumber(): Promise<number>;
    getTransactionCount(blocktag: string | number): Promise<number>;
    getTransactions(address?: Address): Promise<Array<TransactionResponse>>;
    vaultTx(asset: string, amount: ethers.BigNumberish, memo: string): Promise<TransactionResponse>;
    estimateNormalTx(params: NormalTxOpts): Promise<ethers.BigNumberish>;
    normalTx(opts: NormalTxOpts): Promise<TransactionResponse>;
    estimateGasERC20Tx(params: EstimateGasERC20Opts): Promise<ethers.BigNumberish>;
    erc20Tx(opts: Erc20TxOpts): Promise<TransactionResponse>;
}
/**
 * Custom Ethereum client
 */
export default class Client implements EthereumClient {
    private _wallet;
    private _network;
    private _phrase;
    private _provider;
    private _address;
    private _etherscan;
    private _vault;
    constructor(network: Network | undefined, etherscanApiKey: string, providersOpts?: Record<'infura' | 'etherscan', {
        weight: number;
        priority: number;
        apiKey?: any;
    }>, phrase?: Phrase, vault?: string);
    /**
     * Getters
     */
    getAddress(): Address;
    get wallet(): ethers.Wallet;
    get vault(): ethers.Contract | null;
    get network(): Network;
    get provider(): Provider;
    get etherscan(): EtherscanProvider;
    /**
     * changes the wallet eg. when using connect() after init()
     */
    private changeWallet;
    /**
     * changes the provider
     */
    EtherscanProvider(apiKey?: any): Provider;
    /**
     * changes the provider to Infura
     */
    InfuraProvider(apiKey?: any): Provider;
    /**
     * Set's the current network
     */
    setNetwork(network: Network): Network;
    /**
     * Set's the current vault contract
     */
    setVault(vault: string): ethers.Contract;
    /**
     * Generates a new mnemonic / phrase
     */
    static generatePhrase(): Phrase;
    /**
     * Validates a mnemonic phrase
     */
    static validatePhrase(phrase: Phrase): boolean;
    /**
     * Sets a new phrase (Eg. If user wants to change wallet)
     */
    setPhrase(phrase: Phrase): boolean;
    /**
     * Validates an address
     */
    static validateAddress(address: Address): boolean;
    /**
     * Gets the eth balance of an address
     * @todo add start & end block parameters
     */
    getBalance(address?: Address): Promise<ethers.BigNumberish>;
    /**
     * Gets the erc20 asset balance of an address
     */
    getERC20Balance(asset: Address, address?: Address): Promise<ethers.BigNumberish>;
    /**
     * Gets the current block of the network
     */
    getBlockNumber(): Promise<number>;
    /**
     * Returns a Promise that resovles to the number of transactions this account has ever sent (also called the nonce) at the blockTag.
     * @param blocktag A block tag is used to uniquely identify a block’s position in the blockchain:
     * a Number or hex string:
     * Each block has a block number (eg. 42 or "0x2a).
     * “latest”:
     *  The most recently mined block.
     * “pending”:
     *  The block that is currently being mined.
     */
    getTransactionCount(blocktag?: string | number, address?: Address): Promise<number>;
    /**
     * Gets the transaction history of an address.
     */
    getTransactions(address?: Address): Promise<Array<TransactionResponse>>;
    /**
     * Sends a transaction to the vault
     */
    vaultTx(asset: Address, amount: ethers.BigNumberish, memo: string): Promise<TransactionResponse>;
    /**
     * Returns the estimate gas for a normal transaction
     * @param params NormalTxOpts  transaction options
     */
    estimateNormalTx(params: NormalTxOpts): Promise<ethers.BigNumberish>;
    /**
     * Sends a transaction in ether
     */
    normalTx(params: NormalTxOpts): Promise<TransactionResponse>;
    /**
     * Returns a promise with the gas estimate to the function call `transfer` of a contract
     * that follows the ERC20 interfaces
     **/
    estimateGasERC20Tx(params: EstimateGasERC20Opts): Promise<ethers.BigNumberish>;
    /**
     * Returns a promise with the `TransactionResponse` of the erc20 transfer
     */
    erc20Tx(params: Erc20TxOpts): Promise<TransactionResponse>;
}
export { Client };


src/client.ts
import { generateMnemonic, validateMnemonic } from 'bip39'
import { ethers } from 'ethers'
import { Provider, TransactionResponse } from '@ethersproject/abstract-provider'
import { EtherscanProvider, getDefaultProvider, InfuraProvider, FallbackProvider } from '@ethersproject/providers'

import vaultABI from '../data/vault.json'
import erc20ABI from '../data/erc20.json'
import { getAddress, formatEther } from 'ethers/lib/utils'
import { toUtf8String } from '@ethersproject/strings'
import { Network, Address, Phrase, NormalTxOpts, Erc20TxOpts, EstimateGasERC20Opts } from './types'

const ethAddress = '0x0000000000000000000000000000000000000000'

/**
 * Interface for custom Ethereum client
 */
export interface EthereumClient {
  setNetwork(network: Network): Network
  setPhrase(phrase?: string): void
  getAddress(): string
  getBalance(address: Address): Promise<ethers.BigNumberish>
  getBlockNumber(): Promise<number>
  getTransactionCount(blocktag: string | number): Promise<number>
  getTransactions(address?: Address): Promise<Array<TransactionResponse>>
  vaultTx(asset: string, amount: ethers.BigNumberish, memo: string): Promise<TransactionResponse>
  estimateNormalTx(params: NormalTxOpts): Promise<ethers.BigNumberish>
  normalTx(opts: NormalTxOpts): Promise<TransactionResponse>
  estimateGasERC20Tx(params: EstimateGasERC20Opts): Promise<ethers.BigNumberish>
  erc20Tx(opts: Erc20TxOpts): Promise<TransactionResponse>
}

/**
 * Custom Ethereum client
 */
export default class Client implements EthereumClient {
  private _wallet: ethers.Wallet
  private _network: Network
  private _phrase: Phrase
  private _provider: Provider
  private _address: Address
  private _etherscan: EtherscanProvider
  private _vault: ethers.Contract | null = null

  constructor(
    network: Network = Network.TEST,
    etherscanApiKey: string,
    providersOpts?: Record<'infura' | 'etherscan', { weight: number; priority: number; apiKey?: any }>,
    phrase?: Phrase,
    vault?: string,
  ) {
    if (phrase && !validateMnemonic(phrase)) {
      throw new Error('Invalid Phrase')
    } else {
      this._phrase = phrase || generateMnemonic()
      this._network = network
      this._provider = providersOpts
        ? new FallbackProvider([
            {
              provider: new InfuraProvider(network, providersOpts.infura.apiKey),
              weight: providersOpts.infura.weight,
              priority: providersOpts.infura.priority,
            },
            {
              provider: new EtherscanProvider(network, providersOpts.etherscan.apiKey),
              weight: providersOpts.etherscan.weight,
              priority: providersOpts.etherscan.priority,
            },
          ])
        : getDefaultProvider(network)
      this._wallet = ethers.Wallet.fromMnemonic(this._phrase)
      this._address = this._wallet.address
      this._etherscan = new EtherscanProvider(this._network, etherscanApiKey) // for tx history
      if (vault) this.setVault(vault)
      // Connects to the ethereum network with it
      const newWallet = this.wallet.connect(this._provider)
      this.changeWallet(newWallet)
    }
  }

  /**
   * Getters
   */
  getAddress(): Address {
    return this._address
  }

  get wallet(): ethers.Wallet {
    return this._wallet
  }

  get vault(): ethers.Contract | null {
    return this._vault
  }

  get network(): Network {
    return this._network
  }

  get provider(): Provider {
    return this._provider
  }

  // to enable spying on EtherscanProvider.getHistory()
  get etherscan(): EtherscanProvider {
    return this._etherscan
  }

  /**
   * changes the wallet eg. when using connect() after init()
   */
  private changeWallet = (wallet: ethers.Wallet): ethers.Wallet => {
    return (this._wallet = wallet)
  }

  /**
   * changes the provider
   */
  EtherscanProvider(apiKey?: any): Provider {
    const newWallet = this.wallet.connect(new EtherscanProvider(this._network, apiKey))
    this.changeWallet(newWallet)
    return (this._provider = this._wallet.provider)
  }

  /**
   * changes the provider to Infura
   */
  InfuraProvider(apiKey?: any): Provider {
    const newWallet = this.wallet.connect(new InfuraProvider(this._network, apiKey))
    this.changeWallet(newWallet)
    return (this._provider = this._wallet.provider)
  }

  /**
   * Set's the current network
   */
  setNetwork(network: Network): Network {
    if (!network) {
      throw new Error('Wallet must be provided')
    } else {
      this._network = network
      this._provider = getDefaultProvider(network)
      this._etherscan = new EtherscanProvider(network)
      return this._network
    }
  }

  /**
   * Set's the current vault contract
   */
  setVault(vault: string): ethers.Contract {
    if (!vault) {
      throw new Error('Vault address must be provided')
    } else {
      const contract = new ethers.Contract(vault, vaultABI, this._provider)
      this._vault = contract.connect(this.wallet)
      return this._vault
    }
  }

  /**
   * Generates a new mnemonic / phrase
   */
  static generatePhrase(): Phrase {
    return generateMnemonic()
  }

  /**
   * Validates a mnemonic phrase
   */
  static validatePhrase(phrase: Phrase): boolean {
    return validateMnemonic(phrase) ? true : false
  }

  /**
   * Sets a new phrase (Eg. If user wants to change wallet)
   */
  setPhrase(phrase: Phrase): boolean {
    if (!Client.validatePhrase(phrase)) {
      throw new Error('Phrase must be provided')
    } else {
      this._phrase = phrase
      const newWallet = ethers.Wallet.fromMnemonic(phrase)
      this.changeWallet(newWallet)
      return true
    }
  }

  /**
   * Validates an address
   */
  static validateAddress(address: Address): boolean {
    try {
      getAddress(address)
      return true
    } catch (error) {
      return false
    }
  }

  /**
   * Gets the eth balance of an address
   * @todo add start & end block parameters
   */
  async getBalance(address?: Address): Promise<ethers.BigNumberish> {
    if (address && !Client.validateAddress(address)) {
      return Promise.reject('Invalid Address')
    }

    try {
      return await this.wallet.provider.getBalance(address || this._address)
    } catch (error) {
      return Promise.reject(error)
    }
  }

  /**
   * Gets the erc20 asset balance of an address
   */
  async getERC20Balance(asset: Address, address?: Address): Promise<ethers.BigNumberish> {
    if (address && !Client.validateAddress(address)) {
      return Promise.reject('Invalid Address')
    }
    if (!Client.validateAddress(asset)) {
      return Promise.reject('Invalid Asset')
    }
    const contract = new ethers.Contract(asset, erc20ABI, this.wallet)
    const erc20 = contract.connect(this.wallet)
    const erc20Balance = await erc20.functions.balanceOf(address || this._address)
    return erc20Balance
  }

  /**
   * Gets the current block of the network
   */
  async getBlockNumber(): Promise<number> {
    return this.wallet.provider.getBlockNumber()
  }

  /**
   * Returns a Promise that resovles to the number of transactions this account has ever sent (also called the nonce) at the blockTag.
   * @param blocktag A block tag is used to uniquely identify a block’s position in the blockchain:
   * a Number or hex string:
   * Each block has a block number (eg. 42 or "0x2a).
   * “latest”:
   *  The most recently mined block.
   * “pending”:
   *  The block that is currently being mined.
   */
  async getTransactionCount(blocktag: string | number = 'latest', address?: Address): Promise<number> {
    return this.provider.getTransactionCount(address || this.getAddress(), blocktag)
  }

  /**
   * Gets the transaction history of an address.
   */
  async getTransactions(address: Address = this._address): Promise<Array<TransactionResponse>> {
    if (address && !Client.validateAddress(address)) {
      return Promise.reject('Invalid Address')
    } else {
      const transactions = this._etherscan.getHistory(address)
      return transactions
    }
  }

  /**
   * Sends a transaction to the vault
   */
  async vaultTx(asset: Address, amount: ethers.BigNumberish, memo: string): Promise<TransactionResponse> {
    if (!this.vault) {
      return Promise.reject('vault has to be set before sending vault tx')
    }
    if (asset.toString() == ethAddress) {
      return await this.vault.deposit(toUtf8String(memo), { value: amount })
    }
    const contract = new ethers.Contract(asset, erc20ABI, this.provider)
    const erc20 = contract.connect(this.wallet)
    const allowance = await erc20.allowance(this.vault.address, { from: this.wallet.address })
    if (formatEther(allowance) < amount) {
      const approved = await erc20.approve(this.vault.address, amount, { from: this.wallet.address })
      await approved.wait()
    }
    const deposit = await this.vault.deposit(asset, amount, toUtf8String(memo))
    return await deposit.wait()
  }

  /**
   * Returns the estimate gas for a normal transaction
   * @param params NormalTxOpts  transaction options
   */
  async estimateNormalTx(params: NormalTxOpts): Promise<ethers.BigNumberish> {
    const { addressTo, amount, overrides } = params
    const transactionRequest = Object.assign({ to: addressTo, value: amount }, overrides || {})
    return this.wallet.provider.estimateGas(transactionRequest)
  }

  /**
   * Sends a transaction in ether
   */
  async normalTx(params: NormalTxOpts): Promise<TransactionResponse> {
    const { addressTo, amount, overrides } = params
    const transactionRequest = Object.assign({ to: addressTo, value: amount }, overrides || {})
    const transactionResponse = this.wallet.sendTransaction(transactionRequest)
    return transactionResponse
  }

  /**
   * Returns a promise with the gas estimate to the function call `transfer` of a contract
   * that follows the ERC20 interfaces
   **/
  async estimateGasERC20Tx(params: EstimateGasERC20Opts): Promise<ethers.BigNumberish> {
    const { erc20ContractAddress, addressTo, amount } = params
    if (addressTo && !Client.validateAddress(addressTo)) {
      return Promise.reject('Invalid Address')
    }
    if (!Client.validateAddress(erc20ContractAddress)) {
      return Promise.reject('Invalid ERC20 Contract Address')
    }
    const contract = new ethers.Contract(erc20ContractAddress, erc20ABI, this.wallet)
    const erc20 = contract.connect(this.wallet)
    return erc20.estimateGas.transfer(addressTo, amount)
  }

  /**
   * Returns a promise with the `TransactionResponse` of the erc20 transfer
   */
  async erc20Tx(params: Erc20TxOpts): Promise<TransactionResponse> {
    const { erc20ContractAddress, addressTo, amount, overrides } = params
    if (addressTo && !Client.validateAddress(addressTo)) {
      return Promise.reject('Invalid Address')
    }
    if (!Client.validateAddress(erc20ContractAddress)) {
      return Promise.reject('Invalid ERC20 Contract Address')
    }
    const contract = new ethers.Contract(erc20ContractAddress, erc20ABI, this.wallet)
    const erc20 = contract.connect(this.wallet)
    const transactionOverrides = Object.assign({}, overrides || {})
    return erc20.transfer(addressTo, amount, transactionOverrides)
  }
}

export { Client }
